### PR TITLE
REFACTOR: Improve reusability by Decoupling flag modal from flag target.

### PR DESCRIPTION
**Pull request by romanrizzi**
We want chat message's flags to have the same features as topic and posts' flags, but we prefer not having to duplicate core's logic. This PR moves target specific bits to different classes, allowing plugins to flag custom things by providing their own.

### CODE REVIEW

**Review by martin-brennan — APPROVED** [review]
LGTM, I like the "polymorphising" for the flag modals with targets here and it's working well for the chat PR :+1:

File: app/assets/javascripts/discourse/app/lib/flag-targets/flag.js [hunk 1]
```diff
@@ -0,0 +1,57 @@
+import { popupAjaxError } from "discourse/lib/ajax-error";
+
+export default class Flag {
+  flaggingTopic() {
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/lib/flag-targets/flag.js:4]: you use `isXXXX` for editable and just `xxxxx` for flaggingTopic, maybe we should use is in both cases here

File: app/assets/javascripts/discourse/app/lib/flag-targets/flag.js [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    // an instance of ActionSummary
+    let postAction = this.postActionFor(controller);
+
+    let params = controller.get("selected.is_custom_flag")
+      ? { message: this.message }
+      : {};
+
+    if (opts) {
+      params = Object.assign(params, opts);
+    }
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/lib/flag-targets/flag.js:22]: A little bit afraid of the possible side effects of this code. It's possibly mutating stuff without being super explicit about it (this.message for example).

File: app/assets/javascripts/discourse/app/lib/flag-targets/flag.js [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+export default class Flag {
+  flaggingTopic() {
+    return false;
+  }
+
+  isEditable() {
+    return true;
+  }
+
+  createFlag(controller, opts) {
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/lib/flag-targets/flag.js:12]: could probably just be `create`, it's little bit redundant to write `Flag.createFlag()`

File: app/assets/javascripts/discourse/app/lib/flag-targets/topic-flag.js [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    model.get("actions_summary").forEach((a) => {
+      a.flagTopic = model;
+      a.actionType = site.topicFlagTypeById(a.id);
+      lookup.set(a.actionType.get("name_key"), ActionSummary.create(a));
+    });
+    flagController.set("topicActionByName", lookup);
+
+    return site.get("topic_flag_types").filter((item) => {
+      return model.get("actions_summary").some((a) => {
+        return a.id === item.get("id") && a.can_act;
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/lib/flag-targets/topic-flag.js:38]: do we need the get here ?

File: app/assets/javascripts/discourse/app/lib/flag-targets/topic-flag.js [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    return "topic:flag-created";
+  }
+
+  flagsAvailable(flagController, site, model) {
+    let lookup = EmberObject.create();
+
+    model.get("actions_summary").forEach((a) => {
+      a.flagTopic = model;
+      a.actionType = site.topicFlagTypeById(a.id);
+      lookup.set(a.actionType.get("name_key"), ActionSummary.create(a));
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/lib/flag-targets/topic-flag.js:32]: do we need get here?

**Review by jjaffeux — APPROVED** [review]

**Pull request merged**